### PR TITLE
feat: create and edit tasks (#3)

### LINKED ISSUE #3: Create and edit tasks
## User Story
As a store employee, I want to create and edit tasks so that I can track work that needs to be done.

## Acceptance Criteria
- [ ] Form to create a new task (title, description, category, priority, due date)
- [ ] Zod validation on all inputs
- [ ] Edit existing task (same fields)
- [ ] Server-side validation on API route
- [ ] Success/error feedback to user
- [ ] Cancel returns to task list without changes

## Estimation
- **Priority:** P1 (Critical)
- **Size:** 5

### PR DESCRIPTION
**Pull request by defaultdave**
## Summary

- Add full CRUD API for tasks (POST, GET by ID, PUT) with Zod validation and consistent error shapes
- Add reusable TaskForm client component with title, description, category, priority, due date, and status fields
- Add create (/tasks/new), detail (/tasks/[id]), and edit (/tasks/[id]/edit) pages
- Add "New Task" button to task list page
- Comprehensive accessibility: autoFocus, required field indicators, aria attributes, semantic HTML

## Acceptance Criteria

- [x] Form to create a new task (title, description, category, priority, due date)
- [x] Zod validation on all inputs (client-side on blur + submit, server-side on API)
- [x] Edit existing task (same fields + status)
- [x] Server-side validation on API route
- [x] Success/error feedback to user
- [x] Cancel returns to task list without changes

## Test plan

- [x] 38 API route tests (POST, GET, PUT — happy path, validation, 404, JSON errors)
- [x] 27 form component tests (render, validation, blur, aria, submit, cancel, edit pre-fill)
- [x] 31 validator tests (createTaskSchema, updateTaskSchema edge cases)
- [x] TypeScript check passes clean
- [x] ESLint passes clean
- [x] 159/169 tests pass (10 pre-existing DB connectivity failures in schema.test.ts)

## Quality Gate

- [x] TypeScript check
- [x] Lint
- [x] Tests (159 passed)

## Code Review

**Verdict:** REQUEST_CHANGES (applied by team lead)

Findings applied:
- Empty description handling: added Zod preprocess `""` → `null` in both create/update schemas
- API routes: explicit `description: description ?? null` for consistent null storage
- Status field: added `aria-required` attribute
- API routes: added try/catch around Prisma calls for consistent `{ error: string }` on DB errors

## QA

**Verdict:** PASS

All 6 acceptance criteria verified. 96 new tests with adequate coverage across API routes, form component, and validators. Two minor items noted (vacuous redirect test, Prisma error handling) — both addressed by team lead.

## Pipeline

| Stage | Agent | Model | Verdict | Duration | Tokens | Tool Uses |
|-------|-------|-------|---------|----------|--------|-----------|
| Implementation | backend (senior) | sonnet | Complete | 6m 2s | 61,786 | 64 |
| Implementation | frontend (senior) | sonnet | Complete | 44m 54s | 108,764 | 96 |
| Code Review | review | haiku | REQUEST_CHANGES (applied) | 2m 45s | 57,393 | 34 |
| QA Validation | qa | sonnet | PASS | 5m 1s | 76,248 | 46 |

Closes #3

---
<sub>🤖 **senior** (sonnet) | ⏱ 362s | 🔤 61.8k tokens | 🔧 64 tool uses</sub>
<sub>🤖 **senior** (sonnet) | ⏱ 2694s | 🔤 108.8k tokens | 🔧 96 tool uses</sub>
<sub>🤖 **review** (haiku) | ⏱ 165s | 🔤 57.4k tokens | 🔧 34 tool uses</sub>
<sub>🤖 **qa** (sonnet) | ⏱ 301s | 🔤 76.2k tokens | 🔧 46 tool uses</sub>

Generated by `/team 3`

### CODE REVIEW

File: src/lib/validators/task.ts [hunk 1]
```diff
@@ -21,3 +21,70 @@ export const taskQuerySchema = z.object({
...
 
 export type TaskQuery = z.infer<typeof taskQuerySchema>;
+
+/**
+ * Schema for creating a new task.
+ * - title: required, non-empty string
+ * - description: optional string
+ * - categoryId: required, non-empty string
+ * - priority: required enum value
+ * - dueDate: optional ISO 8601 date string (coerced to Date)
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/lib/validators/task.ts]: Fixed in 1f2372d — updated JSDoc to reflect that `description` is nullable (empty string → null via preprocess) and `dueDate` is validated as an ISO 8601 string, not coerced to Date.

File: src/lib/validators/task.ts [hunk 2]
```diff
@@ -21,3 +21,70 @@ export const taskQuerySchema = z.object({
...
+ * The form component converts this to an ISO 8601 datetime before submitting.
+ */
+export const taskFormSchema = z.object({
+  title: z.string().min(1, "Title is required"),
+  description: z.string().optional(),
+  categoryId: z.string().min(1, "Category is required"),
+  priority: priorityEnum,
+  /** YYYY-MM-DD from the HTML date input, or empty string if not set. */
+  dueDate: z.string().optional(),
+  status: statusEnum.optional(),
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/lib/validators/task.ts:87]: Not addressing — `dueDate` in `taskFormSchema` comes exclusively from `<input type="date">`, which browsers constrain to `YYYY-MM-DD` format. Adding a regex validator here would be defensive coding against a scenario that cannot occur through the UI. The server-side `createTaskSchema` already validates the ISO 8601 format as the real boundary check.

File: src/app/api/tasks/[id]/route.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+export async function GET(
+  _request: NextRequest,
+  context: RouteContext
+): Promise<NextResponse> {
+  const { id } = await context.params;
+
+  const task = await prisma.task.findUnique({
+    where: { id },
+    include: taskWithCategoryInclude,
+  });
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/app/api/tasks/[id]/route.ts]: Fixed in 1f2372d — wrapped the `findUnique` in GET in a try/catch with a consistent `{ error: "Failed to fetch task" }` JSON response on DB errors, matching the pattern already used in PUT and POST.

File: src/components/shared/task-form.tsx [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+function isoToDateInputValue(iso: string | null | undefined): string {
+  if (!iso) return "";
+  // iso is like "2026-04-15T00:00:00.000Z"
+  return iso.slice(0, 10);
+}
+
+/** Convert a YYYY-MM-DD string to an ISO 8601 datetime string (midnight UTC). */
+function dateInputToIso(date: string): string | undefined {
+  if (!date) return undefined;
+  return new Date(`${date}T00:00:00.000Z`).toISOString();
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/components/shared/task-form.tsx:51]: Not addressing — `dateInputToIso` is only called with the value from `<input type="date">`, which produces either `""` (handled by the `if (\!date)` guard on line 50) or a valid `YYYY-MM-DD` string. The suggested scenarios (autofill quirks, programmatic input) are hypothetical and would require guarding every internal helper against impossible inputs. The server-side schema provides the real validation boundary.

File: src/components/shared/task-form.tsx [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+            value={values.categoryId}
+            onValueChange={(v) => handleSelectChange("categoryId", v)}
+            disabled={isSubmitting}
+          >
+            <SelectTrigger
+              id="category"
+              aria-required="true"
+              aria-invalid={Boolean(errors.categoryId)}
+              aria-describedby={errors.categoryId ? "category-error" : undefined}
+              className={cn(errors.categoryId && "border-destructive")}
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/components/shared/task-form.tsx:224]: Observation is accurate but no code change needed. The other required fields are Radix Select components, which do not have a meaningful "empty blur" state the way text inputs do — the user either selects a value or the field remains unselected, and submit-time validation catches it. Due Date is optional, so blur validation is not applicable. The PR description wording ("on blur + submit") could be more precise: blur validation applies to text inputs, submit validation covers all fields.

File: src/lib/validators/task.ts [hunk 6]
```diff
@@ -21,3 +21,70 @@ export const taskQuerySchema = z.object({
...
+  status: statusEnum.default("todo"),
+});
+
+export type CreateTaskInput = z.infer<typeof createTaskSchema>;
+
+/**
+ * Schema for updating an existing task.
+ * All fields are required (no partial updates via this schema).
+ * dueDate may be null to clear the date, or a valid ISO 8601 string.
+ * description may be null to clear it, or a string to set it.
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/lib/validators/task.ts]: Fixed in 1f2372d — updated JSDoc to clarify that `title`, `categoryId`, `priority`, and `status` are required, while `description` and `dueDate` are optional/nullable.

File: src/app/api/tasks/[id]/route.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    );
+  }
+
+  const { title, description, categoryId, priority, dueDate, status } =
+    parsed.data;
+
+  const existing = await prisma.task.findUnique({ where: { id } });
+  if (!existing) {
+    return NextResponse.json({ error: "Task not found" }, { status: 404 });
+  }
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/app/api/tasks/[id]/route.ts:62]: Not addressing — catching Prisma's `P2025` error code to distinguish "not found" from other errors is more fragile and less readable than an explicit existence check. The current pattern (`findUnique` → 404, then `update` → catch → 500) is clear and intentional. The extra query is a microsecond-level cost on an indexed primary key lookup. Readability and explicitness are preferred here over saving one query.

File: src/test/components/task-form.test.tsx [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    });
+
+    // The form will fail validation for categoryId (required), so let's set it
+    // If categoryId is empty, the form won't submit to API; let's check redirect
+    // Since categoryId is required, the form will show an error and not call fetch
+    // So this test checks redirect only when the form is fully valid
+    // For simplicity, we check that mockPush is called when API returns ok
+    // To make the form submit, we need to set all required fields
+    // Since this is a unit test, let's just verify behavior: if validation fails, no push
+    // If we add more interactions, let's just verify that a network error leads to error display
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/test/components/task-form.test.tsx]: Fixed in 1f2372d — rewrote this test. It now correctly tests that when required fields are missing (category not selected), validation blocks submission: fetch is not called and no redirect occurs. The test has proper assertions and a name that matches its behavior.

File: src/test/components/task-form.test.tsx [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+      fireEvent.submit(screen.getByRole("form", { name: "Create task" }));
+    });
+
+    await waitFor(() => {
+      expect(screen.getByText("Category is required")).toBeInTheDocument();
+    });
+
+    // fetch should not have been called since validation failed
+    expect(mockFetch).not.toHaveBeenCalled();
+  });
```
[the remark this answers is not in the document]
Comment (defaultdave) [reply src/test/components/task-form.test.tsx]: Fixed in 1f2372d — renamed from "should display server error message on API failure" to "should show category validation error when submitting with title but no category", which accurately describes what the test verifies. Removed the unused mock setup and misleading comments.
